Server handlers

brettz9 opened the pull request:
Note that for the `onerror` test with a bad version (too low), Firefox does not recover. Even if we wrap the `open()` in a try-catch block, Firefox throws an uncatchable error for such `VersionError` errors.

I think therefore, the only way we could work around this would be to always open the current version, reflect on what that version is, close the connection, and then either open the connection if the user-supplied version is good, or immediately reject if it is bad. I don't think it is worth doing this though, just to workaround a bug in Firefox. (I haven't reported the issue yet, however.)

aaronpowell commented:
So what's the overall goal of this PR?

brettz9 commented:
For `onabort` and `onerror`, it allows adding a higher-level listener (as with DOM element bubbling), without needing to add `catch` conditions for every promise that is made:

```js
server.onabort(function (e) {
    displayGenericAbortedMessageToUser();
}).onerror(function (err) {
    displayGenericErrorMessageToUser();
});

server.add(...).then(function () {
    server.query(...).then(function () {};
});
```

The `onversionchange` event handler, on the other hand, can be used to auto-close the database whenever an upgrade is being attempted in say another tab (and which therefore needs for such old connections as in this tab to first be unblocked). It might also notify the user that the page will be refreshed or otherwise updated in order to take advantage of the new schema.

brettz9 commented:
Btw, as you've merged the previous PR already, and I've merged your changes, the diffs here should be much clearer/shorter now...

aaronpowell commented:
Build failed, and this time it's an actual test failure :grinning: 

It might be a PhantomJS problem, but still, it's a failure that's **real** for once!

aaronpowell commented:
I wonder if the events shouldn't operate in the same manner as DOM events, using `addEventListener`/`removeEventListener`, rather than just exposing a single function. This would allow multiple handlers and also cleaner removal.

brettz9 commented:
That's a great idea re: `addEventListener`.

As far as errors, Chrome is ok, but Firefox is still failing for me on the version checks, and there is a new problem I reported in Firefox in the OP of this PR.

aaronpowell commented:
So thinking about this particular failing test - it's most likely a bug in the IDB implementation inside of PhantomJS (well, the QtWebKit it uses) not firing the success event after the versionchange event completes.

This isn't something which we can patch in db.js, it's too low level. Instead for this test what I'd do is move the `done()` call inside the `onversionchange` handler as the test is done when that fires, not when the new version is opened. The promise continuation should still close the db _if it's called_, but it being called is not important to the test passing.

brettz9 commented:
Btw, regarding `addEventListener`/`removeEventListener`, I am thinking that event handlers (similarly to `dbCache`) should be stored separately according to the particular server and schema version name, but not to a table name (so I think the event listener methods will need to be avoided on the store object as with `close`).

A different database should have its own separate events, and if the schema has undergone a version change, changes may be serious enough to warrant new events, but I don't see any particular benefit to tying events to a particular store (it might make a little sense for `onerror` and `onabort` but definitely not for `versionchange`).

What do you think?

aaronpowell commented:
Why not just delegate down to the `addEventListener` on the underlying server/objectStore? Isn't that what you're doing with the events already?

brettz9 commented:
Ok, the `addEventListener`/`removeEventListener` work is done. I've avoided added it to the level of the store, since there is not much advantage to it, and the built-in API doesn't do this anyways. I've changed the API to actually use `addEventListener` as well as using it internally (I had just been using `on*` such as `onerror`, etc.).

As far as the PhantomJS error, I'll take a look at making the fix(es) when Travis is back online.

brettz9 commented:
Hmm..Travis is working again, but apparently I need to get the babel-polyfill added somehow to the grunt file...

brettz9 commented:
Your approach to workaround the versionchange issue worked--it is now passing again!

One other thing...

Since I changed to `addEventListener` not only internally, but also for the API, what do you think of the following API to also utilize `addEventListener` but avoid the bulkier syntax and allow chaining:

```js
server.abort(function (ev) {}).error(function (err) {}).versionchange(function (e) {});
```

brettz9 commented:
Also, the workaround you added in FF doesn't seem to help, so I've removed it

brettz9 commented:
I've added the shorter event syntax--see what you think...

Getting a new error now in the latest builds though and going a bit nuts trying to figure out the problem...

brettz9 commented:
Do you have any tips on how I can debug this?

brettz9 commented:
Problem solved. I assume you'll want a new PR to squash all of the commits (and not sure Github allows squashing to the same branch).

brettz9 commented:
Ok, I've rebased my many recent commits down into one and it is passing Travis. Please review so I can start my next PRs. Thanks!

aaronpowell commented on tests/specs/close-db.js:
```diff
@@ -5,10 +5,17 @@
+            spec.server = undefined;
+            if (!this.dbName) {
+                this.dbName = guid();
+                done();
```
what scenario would result in this happening?

brettz9 replied:
We want to add the `dbName` right away for the first test, but in the case of subsequent tests, we want to first delete the previous database (the old value of `dbName`) and then set the new value of `dbName`, etc.

aaronpowell replied:
Not true - I want a new db name generated in each test, that way we don't have to worry about the previous db being removed, that way failing tests don't cause a chain reaction and make all following tests fail.

brettz9 replied:
I think I ran into a memory issue in some environment or another with the way it had been...

aaronpowell commented on tests/views/layout.jade:
```diff
@@ -3,6 +3,7 @@ html
   head
     meta(charset='utf-8')
     link(type='text/css', rel='stylesheet', href='../node_modules/mocha/mocha.css')
+    link(rel='shortcut icon', href='data:image/x-icon;,', type='image/x-icon')
```
what's this for?

brettz9 replied:
Was just to get FF and Chrome to stop adding a warning to console about no favicon--but this didn't get added to `index.html` for some reason, despite my running the grunt test:local task...

aaronpowell commented on tests/views/index.jade:
```diff
@@ -1,12 +1,14 @@
 extends layout
 
 append scripts
+    script(src='../node_modules/babel-polyfill/dist/polyfill.js')
```
what browser(s) need this?

brettz9 replied:
I didn't run into a problem for Chrome or Firefox--I was adding it:

1. for parity with Karma/PhantomJS (which did need it)
2. any browsers that haven't yet supported it (which apparently may be IE/Edge per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes#Browser_compatibility )

`Array.prototype.includes` is an ES2016 (ES7) feature, so a little bit later coming on...

aaronpowell replied:
Did I merge something with `Array.prototype.includes` without realising? Whoops!

Well yes, that makes sensse then.

Keep in mind that karma doesn't hit the HTML file, so it doesn't need all the polyfills in here too.

brettz9 replied:
I just added `Array.prototype.includes` in this PR--more semantic and clean than `indexOf`.

So do you want the polyfill for the sake of IE/Edge (in case it may work there and somebody actually wants to run tests in it)?

aaronpowell replied:
yeah keep it in. Given I'm an IE/MSEdge MVP I _should_ test in it, but it annoys me that they won't fix the multi-key index bug (well, implement that feature) so I can't add automated tests in Sauce Labs that cover Edge (maybe I should add FF now...)

aaronpowell commented on karma.conf.js:
```diff
@@ -16,6 +22,7 @@ module.exports = function(config) {
     // list of files / patterns to load in the browser
     files: [
       'node_modules/es6-promise/dist/es6-promise.js',
+      'node_modules/babel-polyfill/dist/polyfill.js',
```
does this make the `es6-promise` line above redundant? `core-js` includes ES6 promises: https://github.com/zloirock/core-js#ecmascript-6-promise

aaronpowell commented:
How cool is it that the travis-ci stuff works again :grinning:

brettz9 commented:
Yes, it is cool... The console logging option I added was also very helpful toward my finally debugging it, and the consequent mess I made of many commits forced me to learn how to rebase.

brettz9 commented:
And I removed the redundant es6 promises from package.json and Karma.

I will see about whether I can use `afterEach` to delete the databases to avoid memory issues and start fresh on the `dbName` creation in `beforeEach`

brettz9 commented:
I switched to `afterEach` for deleting... Still curious why there are problems adding those handlers sometimes, but I've removed them and added a comment in the files where they were causing problems.

Anyways, the commit is all rebased and ready for review.

brettz9 commented:
Awesome, thanks!
